Engine/DataFeeds/WorkScheduling/WeightedWorkScheduler.cs

```diff
@@ -0,0 +1,144 @@
+
+        /// <summary>
+        /// Singleton instance
+        /// </summary>
+        public static WeightedWorkScheduler Instance = new WeightedWorkScheduler();
+
+        private WeightedWorkScheduler()
+        {
+            _newWork = new ConcurrentQueue<WorkItem>();
+
+            var work = new List<WorkQueue>();
+            var queueManager = new Thread(() =>
+            {
+                var workersCount = Configuration.Config.GetInt("data-feed-workers-count", Environment.ProcessorCount);
+                MaxWorkWeight = Configuration.Config.GetInt("data-feed-max-work-weight", 400);
+                Logging.Log.Trace($"WeightedWorkScheduler(): will use {workersCount} workers and MaxWorkWeight is {MaxWorkWeight}");
+
+                for (var i = 0; i < workersCount; i++)
+                {
+                    var workQueue = new WorkQueue();
+                    work.Add(workQueue);
+                    var thread = new Thread(() => WorkerThread(workQueue, _newWork))
+                    {
+                        IsBackground = true,
+                        Priority = ThreadPriority.Lowest,
+                        Name = $"WeightedWorkThread{i}"
+                    };
+                    thread.Start();
+                }
+
+                // make sure that the WorkQueue are kept sorted and the weights up to date
+                while (true)
+                {
+                    Thread.Sleep(TimeSpan.FromMilliseconds(1));
+                    foreach (var queue in work)
+                    {
+                        queue.Sort();
+                    }
+                }
+            })
+            {
+                IsBackground = true,
+                Name = "WeightedWorkManager"
+            };
+            queueManager.Start();
+        }
+
+        /// <summary>
+        /// Add a new work item to the queue
+        /// </summary>
+        /// <param name="workFunc">The work function to run</param>
+        /// <param name="weightFunc">The weight function.
+        /// Work will be sorted in ascending order based on this weight</param>
+        public void QueueWork(Func<int, bool> workFunc, Func<int> weightFunc)
+        {
+            _newWork.Enqueue(new WorkItem(workFunc, weightFunc));
+        }
+
+        /// <summary>
+        /// This is the worker thread loop.
+        /// It will first try to take a work item from the new work queue else will check his own queue.
+        /// </summary>
+        private static void WorkerThread(WorkQueue workQueue, ConcurrentQueue<WorkItem> newWork)
+        {
+            while (true)
+            {
+                WorkItem workItem;
+                if (!newWork.TryDequeue(out workItem))
+                {
+                    workItem = workQueue.Get();
+                    if (workItem == null)
+                    {
+                        // no work to do, lets sleep and try again
+                        Thread.Sleep(TimeSpan.FromMilliseconds(1));
```

Comment:
might benefit from making this a spinway  ( see `ISleepStrategy` for abstracting this ) -- I think it's _very_ likely that new jobs could be added faster than once a millisecond -- also -- the OS doesn't guarantee you'll wake up in 1ms, it just guarantees you won't wake up _before_ 1ms. I would like to see benchmarks for the following:
``` csharp
// 1. guarantees the thread does not start executing until a full ms has passed
Thread.Sleep(1) (also, use 1 instead of timespan, no need to create a timespan each time)

// 2. yield to other processes w/ same or higher priority (unlikely since elevated priority configured
Thread.Sleep(0)

// 3. forcefully yield to other processes waiting to run
Thread.Yield();

// 4. Uses spin wait until a predefined max wait before switching to event wait handles
spin = new SpinWait();
spin.SpinOnce();

// 5. Use ManualResetEventSlim to signal when items are added to stop waiting
sync = new ManualResetEventSlim();
// use sync.Reset() after dequeueing everything
// use sync.Set() after enqueueing something
// use sync.WaitOne(1) to wait for work, after 1 ms check the queue anyway
```

My hunch is that the `Thread.Yield()` and `Thread.Sleep(0)` likely cause a degree of thread starvation since the priorities are jacked up. It also has the downside of pegging the CPU. I think the best option here is likely to be the `ManualResetEventSlim`. Internally he does a spin wait, not sure for how long, but after the spin wait he goes to a different wait. Using spin wait means the absolute fastest possible signaling from item added to pulling it out of the queue, without pegging the CPU of course.

Reply:
Note that new jobs get into their own `newWork` queue which is _shared_ among all the worker threads, so the 1ms delay will probably get significantly reduced since it will be hard for all the workers to be sleeping at the same time. Played around a bit with `Thread.Yield()` but saw no improvements. I agree we don't want to spin wait on the CPU since it would be really expensive.
I think maybe an event could also be a possible solution but it would get ugly fast, I'll give it a shot 👍: We would have to make the 'ManagerThread' notify each thread when there is a job available -> checking the weights and detecting they are bellow the threshold + a notification system for new work items coming in too.

Reply:
Implemented in commit 2: added an event for each work queue and an event for the general new work queue 👍 . Added test case 8, `HistoryRequestBenchmark`, where it can be seen that commit 1 was adding an overhead, now fixed.